itests/org.openhab.core.io.net.tests/src/main/java/org/openhab/core/io/net/tests/internal/TestServer.java

```diff
@@ -61,6 +65,17 @@ public void startServer() throws Exception {
         connector.setIdleTimeout(timeout);
         server.addConnector(connector);
 
+        server.addLifeCycleListener(new LifeCycle.Listener() {
+            @Override
+            public void lifeCycleStarted(@Nullable LifeCycle event) {
+                if (event != null && event.isStarted()) {
+                    isStarted.complete(true);
+                } else {
+                    isStarted.completeExceptionally(new IllegalStateException("Server not started"));
+                }
+            }
+        });
+
         server.start();
```

Comment:
`Server.start()` won't return until the server is started.
This `LifeCycleListener` isn't doing anything.

What you need to be paying attention to is the WebApp deployed / started.

Reply:
> What you need to be paying attention to is the WebApp deployed / started.

@joakime thanks for the feedback. I guess I misunderstood that the lifecycle notification is only sent after everything in the server and all its nested connectors and handlers etc. are running. So can you please be more specific about 'checking if the WebApp is running'? In this case the 'applications' are servlets that are nested in a handler which is nested in the server (which is why I was looking at the lifecycle notification test). i.e. in other words, is it possible that something might be 'started' but eventually not actually 'running' resp. not yet processing incoming data??

Reply:
.. or to put it another way: How can it be that we get 'connection refused' after the 'start()` method has returned? The server is obviously running, since otherwise the connection would passively timeout rather than being actively refused..

Reply:
That would mean that the test is starting before the `Server.start()` has completed.
Should be a simple enough thing to double check.

Do you add your `Connectors` after `Server.start()`? (if yes, then you need to wait for those to reach "started" in their LifeCycle)

In Jetty, we test the real Jetty `Server` everywhere in our own unit tests.
We rely on the `LifeCycle.start()` behavior to behave consistently and predictably for this to work. (and it has, for 22 years)

If we deploy a webapp after the `Server.start()`, then we have to test for if that webapp is started before we can send requests to that webapp.

If we have a `DeploymentManager` doing the deployment, then we can add a `AppLifeCycle.Binding` to a binding target of `"deployed"` to trigger our tests to start making requests to that webapp.

If we add the `WebAppContext` or `ServletContextHandler` or `ContextHandler` to the server handler tree manually (not via a DeploymentManager), then we can either use a LifeCycleListener to tell us about the `lifeCycleStarted` event for that webapp, or we can periodically test the `LifeCycle.isStarted()` method against that webapp to know if it's started yet. (We usually use the LifeCycleListener on webapp, and a `java.util.concurrent.CountDownLatch` to trigger off the started event in our test cases)

Reply:
Hmm. All connectors and servlets are added to the server before server.start() gets called. And (at least according to the itest @BeforeAll annotation) the client tests (annotated @Test) can (in theory) only start after the server.start() method (in @BeforeAll) has completed. However in reality sometimes we see a 'connection refused' result from the first test. .. I suppose there could be a threading issue whereby an @Test method is called on one thread, before the @BeforeAll method on another thread has fully completed. Or domething like that. I will try adding a synchronized lock to prevent that.

Reply:
It could also be a test that's still running when your `@AfterAll` runs and shuts down/stops the Server.

Reply:
^
Good point.

Reply:
@joakime many thanks for the guidance. Through it I have learned a lot about Jetty (albeit not even approaching your own knowledge). I added some Completable's to ensure correct pipelining of @BeforeAll => @Test => @AfterAll methods.